### PR TITLE
Add search space class for BayesOpt

### PR DESCRIPTION
**Pull request by Thomas-Christie**
Added an `AbstractSearchSpace` abstract base class for the BayesOpt library, as well as a concrete implementation in the form of a `ContinuousSearchSpace` class. Tests have also been added.

## Type of changes

- [ ] Bug fix
- [X] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [X] I've added tests for new code.
- [X] I've added docstrings for the new code.

## Description

Starting to add fully tested Bayesian Optimisation functionality. This first PR adds a class for defining (and sampling from) search spaces, over which optimisation is performed.

Issue Number: N/A

### CODE REVIEW

File: gpjax/bayes_opt/search_space.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+
+@dataclass
+class ContinuousSearchSpace(AbstractSearchSpace):
+    """The `ContinuousSearchSpace` class is used to bound the domain of continuous real functions of dimension $`D`$."""
+
+    lower_bounds: Float[Array, " D"]
+    upper_bounds: Float[Array, " D"]
+
+    def __post_init__(self):
```
Comment (henrymoss) [inline gpjax/bayes_opt/search_space.py:64]: worth checking that same dtype for lower and upper?

File: gpjax/bayes_opt/search_space.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        if self.lower_bounds.shape != self.upper_bounds.shape:
+            raise ValueError("Lower and upper bounds must have the same shape.")
+        if self.lower_bounds.shape[0] == 0:
+            raise ValueError("Lower and upper bounds cannot be empty")
+        if not (self.lower_bounds < self.upper_bounds).all():
```
Comment (henrymoss) [inline gpjax/bayes_opt/search_space.py:69]: I think we should allow for some of the bounds to be equal, e.g. if we have a particular dim that we wanna keep fixed

File: gpjax/bayes_opt/search_space.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            raise ValueError("Lower bounds must be less than upper bounds.")
+
+    @property
+    def dimensionality(self) -> int:
+        return self.lower_bounds.shape[0]
+
+    def sample(self, num_points: int, key: KeyArray) -> Float[Array, "N D"]:
+        """Sample points from the search space using a Halton sequence.
+
+        Note that a single key generates a single sequence so, for instance, if one
```
Comment (henrymoss) [inline gpjax/bayes_opt/search_space.py:79]: im not sure this is true with this implementation? By default randomized=True is the tfp.mcmc.sample_halton_sequence.

I thought that re-jigged stuff?

https://www.tensorflow.org/probability/api_docs/python/tfp/mcmc/sample_halton_sequence

**Review by henrymoss — APPROVED** [review]
Nice work. Just a couple comments

File: gpjax/bayes_opt/search_space.py (hunk 1 above)
Comment (Thomas-Christie) [reply gpjax/bayes_opt/search_space.py:64]: Will do (and add test)

File: gpjax/bayes_opt/search_space.py (hunk 3 above)
Comment (Thomas-Christie) [reply gpjax/bayes_opt/search_space.py:79]: Have tested and different seeds (even if close in value) produce sufficiently different samples (i.e. the seed isn't being treated as some `skip` index). Did notice, however, that with the same key (and different `num_points`), it isn't guaranteed that the samples form prefixes of each other. It seems that the randomisation resets on each power of 2 such that, for instance, sequences of length 16-31 will have the same prefix. Have decided that this is fine (as the main guarantee we want is different keys => different samples).

**Comment by henrymoss** [comment]
@thomaspinder we have a working full version of BO that we like now in another branch. I jsut suggested to Thomas that he chopped it up into a few manageable bits, e.g. just the search space PR

**Comment by Thomas-Christie** [comment]
Yes @thomaspinder, @henrymoss and I decided we'd split up PR https://github.com/JaxGaussianProcesses/GPJax/pull/338 into a few fully tested PRs for ease of review. I'm happy to just merge into `main` to simplify things, as these PRs should all be getting added within the next week or two, but am happy to merge onto a separate BO branch if you'd prefer us not to merge onto `main` for now.

**Comment by Thomas-Christie** [comment]
Will merge this PR onto the `tchristie/bo` branch.

**Pull request merged**